### DISCUSSION TITLE
Does ProjectPackage example in Micosoft Doc work with Community Toolkit

### DISCUSSION THREAD

**Discussion by coleblarson** [VsixCommunity/Community.VisualStudio.Toolkit#288] [Q&A]
Is the example here : https://docs.microsoft.com/en-us/visualstudio/extensibility/creating-a-basic-project-system-part-1?view=vs-2022

Compatible with a toolkit derived project ...  if not, is an example showing how to build a project package with the Community Toolkit available somewhere?   (the sample says you have to derive a project package class from ProjectPackage)

**Comment by reduckted** [comment]
I f I understand that guide correctly, the `ProjectPackage` they are referring to is [this one](https://github.com/tunnelvisionlabs/MPFProj10/blob/master/Dev10/Src/CSharp/ProjectPackage.cs#L62), which you need to download the code for and include in your project, rather than referencing an assembly. It's not an `AsyncPackage`, so as it is, I'm not sure Visual Studio 2022 will even load that package (I could be wrong, but I thought all packages now need to inherit `AsyncPackage` instead of the older synchronous `Package` class).

However, since it looks like you need to include the source code directly instead of referencing a library, yes it _could_ be compatible because you could change that code to inherit from `ToolkitPackage` instead of `Package`.

_But_, I would follow the suggestion at the top of that article instead:

> If you need to extend Visual Studio with a custom project type, we strongly recommend leveraging the [Visual Studio project system](https://github.com/Microsoft/VSProjectSystem) (VSPS) which has a number of advantages over building a project system from scratch:
> 
> * Easier onboarding. Even a basic project system requires tens of thousands of lines of code. Leveraging VSPS reduces the onboarding cost to a few clicks before you are ready to customize it to your needs.
> * Easier maintenance. By leveraging VSPS, you only need to maintain your own scenarios. We handle the upkeep of all of the project system infrastructure.

**Comment by RobertvanderHulst** [comment]
We have created a project system based on the MPF code and enhanced with some additions from the project system that the Wix team used (mainly the support for "Show all Files").
We have also looked at the VSProject system (AKA CPS) , but we could not get that system to work reliably for Windows Forms and WPF style projects.
If you want to see what we have done: https://github.com/X-Sharp/XSharpPublic/tree/main/VisualStudio. The ProjectBase folder mainly contains the MPF code. The ProjectPackage folder has our own code. There is also a ProjectPackageCPS folder which contains our attempt to build a project system on top of CPS.

We have modified the MPF code so the base class of the packages is now the ToolkitPackage.
We have also followed the threading advises and added code to switch to the main thread where needed.
We have also added support this year for PackageReferences.

Our project system works with VS 2017 & VS 2019 in one package and VS 2022 in the other. We looked at the shared project approach but found it easier to just have 2 csproj files in the same folder with different output folder names.

The only disadvantage of our current solution is that it still uses the "old style" Add Reference dialogs. We looked at the newer style but had some problems to get that to work reliably. I should probably look at that again.
